File: problem1/PathSum2.java
// Time Complexity : O(n)
// Space Complexity : O(h)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
package problem1;

import java.util.ArrayList;
import java.util.List;

public class PathSum2 {
	public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
		List<List<Integer>> ans = new ArrayList<>();
		if (root == null) {
			return ans;
		}

		helper(root, targetSum, new ArrayList<Integer>(), ans);
		return ans;
	}

	public void helper(TreeNode root, int targetSum, List<Integer> pathVal, List<List<Integer>> ans) {
		if (root == null) {
			return;
		}

		pathVal.add(root.val);

		if (targetSum == root.val && root.left == null && root.right == null) {
			ans.add(new ArrayList<>(pathVal));
		} else {
			helper(root.left, targetSum - root.val, pathVal, ans);
			helper(root.right, targetSum - root.val, pathVal, ans);
		}
		pathVal.remove(pathVal.size() - 1);
	}

	public static void main(String[] args) {
		PathSum2 obj = new PathSum2();

		TreeNode root = new TreeNode(5);
		root.left = new TreeNode(4);
		root.right = new TreeNode(8);
		root.left.left = new TreeNode(11);
		root.left.left.left = new TreeNode(7);
		root.left.left.right = new TreeNode(2);
		root.right.left = new TreeNode(13);
		root.right.right = new TreeNode(4);
		root.right.right.left = new TreeNode(5);
		root.right.right.right = new TreeNode(1);

		int target = 22;
		System.out.println("Paths with target " + 22 + ": ");
		List<List<Integer>> paths = obj.pathSum(root, target);
		for (List<Integer> path : paths) {
			for (Integer node : path) {
				System.out.print(node + " ");
			}
			System.out.println();
		}
	}
}

File: pathSum2.js
// Time Complexity : O(n)
// Space Complexity : O(h) - worst case: O(n) if we come across a skewed tree
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

// Optimized Approach: Instead of creating a new array and passing that in each time
// we call the dfs function, we now use the same array and pop stuff out as we unroll the stack.
// As for pushing an array to the results array, we send a copy of the existing array so that the
// array that is referenced can become empty at the end once the stack becomes empty.
var pathSum = function (root, targetSum) {
  if (!root) return [];
  const resultsArr = [];

  const dfs = (root, cSum, valArr) => {
    // base condition
    if (root === null) return;

    // logic
    valArr.push(root.val);
    cSum = cSum + root.val;
    dfs(root.left, cSum, valArr);
    dfs(root.right, cSum, valArr);
    if (root.left === null && root.right === null) {
      if (cSum === targetSum) {
        resultsArr.push([...valArr]);
      }
    }
    valArr.pop();
  };

  dfs(root, 0, []);
  return resultsArr;
};

// Slower approach since we're utilizing a new array each time, instead of
// mutating the existing one. This will return an O(n)^2 for both TC & SC.
var pathSum = function (root, targetSum) {
  if (!root) return [];
  const resultsArr = [];

  const dfs = (root, cSum, valArr) => {
    // base condition
    if (root === null) return;

    // logic
    valArr.push(root.val);
    cSum = cSum + root.val;
    dfs(root.left, cSum, [...valArr]);
    dfs(root.right, cSum, [...valArr]);
    if (root.left === null && root.right === null) {
      if (cSum === targetSum) {
        resultsArr.push(valArr);
      }
    }
  };

  dfs(root, 0, []);
  return resultsArr;
};
